### PR TITLE
Add `preact/compat` support to `@astrojs/preact`

### PR DESCRIPTION
**Pull request by delucis**
## Changes

This branch tries to add a `compat` configuration option to the `@astrojs/preact` integration based on the the [old `preact/compat` renderer](https://github.com/withastro/astro/blob/f892aeb52f5a93d81a68d9833eb26bedd06aa2f0/packages/renderers/renderer-preact/compat/index.js).

I don’t think this works yet — trying with a local copy of this integration on Stackblitz still produces a `Cannot find module 'react'` error: https://stackblitz.com/edit/github-6sydzv?file=integrations/preact.mjs

I couldn't find any documentation or tests for the old renderer, so I'm not totally clear how it was used. I imagine you always need `@astrojs/preact` as well as compat, so thought this API would be nice:

```js
defineConfig({
  integrations: [
    preact({ compat: true }),
  ],
});
```

If that could work, I think it would be the simplest Preact compat support around?

## Testing

Nate added an E2E test for the compat option

## Docs

Chris updated the integration README to follow our new template and add details about the new `compat` config option.

### CODE REVIEW

File: packages/integrations/preact/src/index.ts [hunk 1]
```diff
@@ -33,16 +51,30 @@ function getViteConfiguration() {
...
+      },
+      dedupe: ['react', 'react-dom'],
+    };
+  }
+	
+	return viteConfig
 }
 
-export default function (): AstroIntegration {
+export default function ({ compat }: { compat?: boolean } = {}): AstroIntegration {
```
Comment (natemoo-re) [inline packages/integrations/preact/src/index.ts:69]: I really like this adjustment to the API! Super clean 👍🏻

File: packages/integrations/preact/src/index.ts [hunk 2]
```diff
@@ -18,8 +18,26 @@ function getRenderer() {
 	};
 }
 
-function getViteConfiguration() {
+function getCompatRenderer(): AstroRenderer {
 	return {
+		name: '@astrojs/preact/compat',
+		clientEntrypoint: '@astrojs/preact/client.js',
+		serverEntrypoint: '@astrojs/preact/server.js',
```
Comment (natemoo-re) [inline packages/integrations/preact/src/index.ts:25]: I wonder if these still work or if we need a separate entrypoint that references `preact/compat`?

**Review by natemoo-re — COMMENTED** [review]
Looking good! 

I bet the issue with any errors like `"react" not found` happens when any dependencies are treated as `vite.ssr.external`—they'll use standard Node resolution rather than Vite's resolution. So in that case, the error is correct... `react` cannot be found since it's not installed.

The fix is possibly to set `vite.ssr.noExternal` to `true`, ensuring that all dependencies are built.

File: packages/integrations/preact/src/index.ts (hunk 2 above)
Comment (delucis) [reply packages/integrations/preact/src/index.ts:25]: They certainly _used_ to be shared in the old compat integration, so I’ll try your `ssr.noExternal` suggestions first and see how I get on. Thanks Nate!

**Comment by natemoo-re** [comment]
@delucis I was able to get things mostly working by using manual aliases via `babel-plugin-module-resolver`. Not sure _why_ it works when the alternative (plain Vite) doesn't, but it seems to be effective. Let me know if you run into anything else!

**Comment by delucis** [comment]
Amazing work @natemoo-re!!! 🥳 

This is very close to perfect. Here’s what works and doesn’t work now:

|                   | Preact component | React component | Preact importing React |
| ----------------- | :--------------: | :-------------: | :--------------------: |
| `build` render    |        ✅        |       ✅        |           ✅           |
| `build` hydration |        ✅        |       ✅        |           ✅           |
| `dev` render      |        ✅        |       ✅        |           ✅           |
| `dev` hydration   |        ✅        |       ❌        |           ❌           |

In `dev`, when the page tries  to hydrate a React component or Preact component that uses a React component, I get the following errors in the browser:

image

Weirder is that the same behaviour happens on a page with just a Preact component if you visit it _after_ having visited a page with a React component. It’s almost like that 504 Error crashes some internal or causes Vite to cache some bad code? Not sure.

I’ll also stress — in case it’s important — that I’m testing this by copy-pasting the changes as a local integration in my test project. It’s on Stackblitz here: https://stackblitz.com/edit/github-6sydzv?file=integrations/preact.mjs

**Review by natemoo-re — APPROVED** [review]
LGTM whenever this is ready for review! It'd be nice to add some docs about this to the README.

**Comment by sarah11918** [comment]
Giving a README LGTM!

File: packages/integrations/preact/README.md [hunk 3]
```diff
@@ -2,57 +2,80 @@
...
-export default {
+export default defineConfig({
   // ...
   integrations: [preact()],
-}
+});
 ```
 
-## Getting started
+Then, restart the dev server.
```
Comment (kevinzunigacuellar) [inline packages/integrations/preact/README.md:75]:
```suggestion
Finally, restart the dev server.
```

**Comment by kevinzunigacuellar** [comment]
LGTM 🚀

**Pull request merged**